text/0253-cdk-metadata-v2.md

```diff
@@ -0,0 +1,488 @@
+with bits anymore once the data hits an AWS datacenter.
+
+# Extensions
+
+We will be able more information into additional fields of the metadata resource.
+Size of template, number of resources are obvious candidates.
+
+We will be able to support non-jsii languages by adding metadata-specific APIs
+to the CDK.
+
+The jsii RTTI mechanism can be extended in the future as well, though those extensions
+are out of scope of this RFC.
+
+If we want to change the encoding of the payload in any way, we can.
+
+If the payload grows too large and changing the encoding scheme won't be
+enough to reduce it, we can transport and reference the payload as an asset.
+Though, that would require us to gain read access to a customer's S3 bucket
+from our account, which has security implications; it would also require a
+customer to have a bootstrapped account for *every* deployment, even small
+ones (though we could lift *that* limitation by hosting bootstrap buckets on behalf
+of users).
+
+# Unresolved questions
+
+- How acceptable is the giant base64-encoded blob in the template? If we want
+  to cut down on it, what clever strategies can we come up with?
+
+# Implementation Plan
+
+This can be implemented in v1 already, no need to wait for v2.
+
+* Implement RTTI support in in jsii.
+* Inform dependent teams that they need to upgrade to the jsii version that supports
+  RTTI.
+* Update metadata service backend to accept new `Analytics` field. This is where
+  decoding should happen back to a flat list again. Metadata service should extract
+  packages from the construct list to continue emitting the old information as well.
+* Update canary to exercise new field.
+* Update metadata reporting job to report on both new and old information.
+* Update metadata resource construction implementation in CDK to switch to RTTI
+  implementation, and start emitting `Analytics` field.
+* Verify metrics are coming in
+
+# Appendix 1: jsii Run-Time Type Information
+
+We will start by having the jsii compiler add FQN information to every
+JavaScript class it emits.
+
+Effectively, every class will be emitted as this in the `.js` file (the
+`.d.ts` file remains unchanged).
+
+```js
+// This will be added to the top of every source file
+const vfqnSym = Symbol.for('jsii.vfqn');
+
+export class SomeClass {
+  // Every class gets an additional field added to it
+  private static [vfqnSym] = 'module.SomeClass@1.2.3';
+
+  // Regular class emits here...
+}
+```
+
+We will introduce a new jsii package to access this information, called `jsii-rtti`.
+
+It will have the following API:
+
+```ts
+class TypeInformation {
+  /**
+   * Get type information based on an object instance
+   */
+  public static forObject(object: any): TypeInformationForObject | undefined;
+
+  /** Type FQN */
+  public readonly fqn: string;
+
+  /** Module version */
+  public readonly version: string;
+
+  /** Base class type information */
+  public readonly baseClass: TypeInformation | undefined;
```

Comment:
Given this format, I actually think we'll be significantly better of modeling the `[Symbol.for('jsii.vfqn')]` property, so we save having to parse it out (i.e: actually set it to some object such as `{ fqn, version }`).

Reply:
Sure